### ISSUE TITLE
`brew deps --installed` Lists Secondary Dependencies of Uninstalled Primary Dependencies

### ISSUE THREAD

**Issue by the-real-tokai** [Homebrew/homebrew#44386]
In some cases `brew deps --installed <formula>` lists secondary dependencies (via recursion) of primary dependencies (level 1) that are not installed on the system.

## Example

```
$ brew info links
links: stable 2.11 (bottled)
Lynx-like WWW browser that supports tables, menus, etc.
http://links.twibright.com/
/usr/local/homebrew/Cellar/links/2.11 (18 files, 1.4M) *
  Built from source with: --with-jpeg --with-libtiff
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/links.rb
==> Dependencies
Build: pkg-config ✔
Recommended: openssl ✔
Optional: libressl ✘, libtiff ✔, jpeg ✔, librsvg ✘
…

$ brew deps --installed --1 links
jpeg
libtiff
openssl
pkg-config
```

And the secondary dependencies of those 4 primary dependencies:
```
$ brew deps jpeg
$ brew deps libtiff
jpeg
$ brew deps openssl
makedepend
pkg-config
$ brew deps pkg-config
```

So all in all we only should have 5 (installed) dependencies in total in this case: *jpeg*, *libtiff*, *openssl*, *pkg-config*, and *makedepend*, but `brew deps --installed links` lists a whole plethora of other dependencies:

```
$ brew deps --installed links
autoconf
automake
cairo
fontconfig
freetype
gettext
glib
gobject-introspection
harfbuzz
icu4c
jasper
jpeg
libffi
libpng
libtiff
libtool
makedepend
openssl
pango
pixman
pkg-config
xz
```

What is going on here? :see_no_evil: 

While it's not listed in the above output —as they are not installed on my system— *links* has optional primary (level 1) dependencies. In the example case the optional *librsvg* drags in all those secondary dependencies to the output.

Those secondary sub-dependencies are installed on my system, but the "connecting" formula *librsvg* is **not**, so those dependencies have no current relevance to *links* and therefore shouldn't be listed as dependency for *links* to avoid user confusion during system maintenance.

I'm not entirely sure if `--installed` is perhaps supposed to work like that, but it certainly is confusing to me. Perhaps needs some kind of `--linked` or `--used` kind of switch?

*This is somewhat related to #43956 (but a different problem.)*

**Comment by tdsmith** [comment]
From `man brew`: `If --installed is passed, show dependencies for all installed formulae.`

`--installed` is a noun, not an adverb; it's being ignored here.

**Comment by tdsmith** [comment]
Well, the output of `comm -3 <(brew deps imagemagick) <(brew deps --installed imagemagick)` isn't empty, so --installed is doing _something_, but it's doing something undefined.

**Comment by tdsmith** [comment]
Oh, of course: it's showing you the intersection of `brew deps --installed` and `brew deps links`, which is sensible.

**Comment by the-real-tokai** [comment]
Not sure, but maybe a full tree helps to demonstrate the problem I have with it? :sweat_smile: 


```
links
├── pkg-config
├── jpeg
├── libtiff
|   └── jpeg
└── openssl
|   └── makedepend
|       └── pkg-config
├── (libressl)               Note: optional, not installed, not listed (OK)
└── (librsvg)                Note: optional, not installed, not listed (OK), but…
    |                            all its dependencies are listed (not OK):
	├── xz
	├── pkg-config
	├── cairo
	|   ├── xz
	|   ├── pkg-config
	|   ├── freetype
	|   |   └── libpng
	|   |       └── xz
	|   ├── fontconfig
	|   |   ├── pkg-config
	|   |   └── freetype
	|   |       └── libpng
	|   |           └── xz
	|   ├── libpng
	|   |   └── xz
	|   ├── pixman
	|   |   └── pkg-config
	|   └── glib
	|       ├── xz
	|       ├── pkg-config
	|       ├── gettext
	|       |   └── xz
	|       └── libffi
	├── gdk-pixbuf
	|   ├── xz
	|   ├── pkg-config
	|   ├── glib
	|   |   ├── xz
	|   |   ├── pkg-config
	|   |   ├── gettext
	|   |   |   └── xz
	|   |   └── libffi
	|   ├── jpeg
	|   ├── libtiff
	|   |   └── jpeg
	|   ├── libpng
	|   |   └── xz
	|   └── gobject-introspection
	|       └── :git
	|       ├── xz
	|       ├── pkg-config
	|       ├── glib
	|       |   ├── xz
	|       |   ├── pkg-config
	|       |   ├── gettext
	|       |   |   └── xz
	|       |   └── libffi
	|       └── libffi
	├── glib
	|   ├── xz
	|   ├── pkg-config
	|   ├── gettext
	|   |   └── xz
	|   └── libffi
	├── libcroco
	|   ├── xz
	|   ├── pkg-config
	|   ├── intltool
	|   └── glib
	|       ├── xz
	|       ├── pkg-config
	|       ├── gettext
	|       |   └── xz
	|       └── libffi
	└── pango
	    ├── xz
	    ├── pkg-config
	    ├── …
	    …
```

Note how most secondary dependencies are dependencies of *librsvg* or one of its sub-dependencies. Even *librsvg* is not installed on my system the command `brew deps --installed links` lists all the sub-dependencies of librsvg that I have installed (it doesn't list librsvg itself, as it is not installed). 

When I type the command `brew deps --installed links` I try to find out what formulas are used by *links*, e.g. which libraries are "linked" (no pun intended) into the binary or which other files (not in the links directory) are used/ required. This is not possible currently. The output of the command suggests that directly or indirectly all those dependencies (*glib*, *gobject-introspection*, etc.) are used/ required.

F.ex. assuming I would like to kill *glib* from my system. The output is telling me that I would need to rebuild/ reinstall the *links* formula afterwards. Which is not the case (which is not obvious, obviously. :smile:) )

**Comment by tdsmith** [comment]
I agree that the thing that you want is not currently provided directly by a `brew` subcommand, but this behavior of `deps` is correctly documented.

**Comment by tdsmith** [comment]
Closing this as not a bug.

**Comment by the-real-tokai** [comment]
Well… how about an option (as I suggested) which changes this behaviour ("--do-not-recurse-over-uninstalled-deps" or shorter "--linked" or "--used")? I did not claim that it was a "bug" necessarily (could have been classified as "feature"). :smile:

Right now, for me as user, this is almost useless or at least very confusing when it reports things that way. When I pass the "--installed" modificator then I expect it to report the situation on disk and not some theoretical formula interconnection. And `man brew` doesn't explain it otherwise to me.

For the reverse case it is true too, of course:

```
$ brew uses --recursive --devel --HEAD --installed pango
doxygen	     fontforge	  graphviz     imagemagick  links	 mtr
```

(reports `links` f.ex. even there's no connection between those both on disk)
